File: frontend/src/lib/api/finance.ts
// Manner_Web - 可以在 Linux 系统上运行的企业管理系统
// Copyright (C) 2026 Linux-System-0(Github) / 一架在Linux上起飞的A320(Bilibili) <ls0_1@qq.com>
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

import { client } from './client'
import type {
  ApiResponse,
  Budget,
  DepartmentReportRow,
  Invoice,
  PaginatedData,
  Payment,
  Reimbursement,
  ReimbursementDetail,
  ReimbursementStatus,
  ReportSummary,
  TrendRow,
} from '@/types'

// ---- 报销单 ----

export interface ReimbursementQuery {
  page?: number
  page_size?: number
  status?: ReimbursementStatus | ''
  keyword?: string
  department_id?: string
}

export async function getReimbursements(
  params: ReimbursementQuery = {},
): Promise<ApiResponse<PaginatedData<Reimbursement>>> {
  const res = await client.get<PaginatedData<Reimbursement>>(
    '/finance/reimbursements',
    params as Record<string, never>,
  )
  return res
}

export async function getReimbursement(
  id: string,
): Promise<ApiResponse<ReimbursementDetail>> {
  const res = await client.get<ReimbursementDetail>(`/finance/reimbursements/${id}`)
  return res
}

export async function createReimbursement(data: {
  title: string
  category: string
  amount: number
  reason?: string
  invoice_ids?: string[]
}): Promise<ApiResponse<{ id: string }>> {
  const res = await client.post<{ id: string }>('/finance/reimbursements', data)
  return res
}

export async function updateReimbursement(
  id: string,
  data: {
    title?: string
    category?: string
    amount?: number
    reason?: string | null
    invoice_ids?: string[]
  },
): Promise<ApiResponse<null>> {
  const res = await client.put<null>(`/finance/reimbursements/${id}`, data)
  return res
}

export async function deleteReimbursement(id: string): Promise<ApiResponse<null>> {
  const res = await client.delete<null>(`/finance/reimbursements/${id}`)
  return res
}

export async function approveReimbursement(
  id: string,
  action: 'approve' | 'reject',
  comment?: string,
): Promise<ApiResponse<null>> {
  const res = await client.post<null>(`/finance/reimbursements/${id}/approve`, {
    action,
    comment,
  })
  return res
}

export async function reviewReimbursement(
  id: string,
  action: 'approve' | 'reject',
  comment?: string,
): Promise<ApiResponse<null>> {
  const res = await client.post<null>(`/finance/reimbursements/${id}/review`, {
    action,
    comment,
  })
  return res
}

export async function payReimbursement(id: string): Promise<ApiResponse<null>> {
  const res = await client.post<null>(`/finance/reimbursements/${id}/pay`)
  return res
}

export async function withdrawReimbursement(id: string): Promise<ApiResponse<null>> {
  const res = await client.post<null>(`/finance/reimbursements/${id}/withdraw`)
  return res
}

// ---- 发票 ----

export interface InvoiceQuery {
  page?: number
  page_size?: number
  keyword?: string
  status?: string
}

export async function getInvoices(
  params: InvoiceQuery = {},
): Promise<ApiResponse<PaginatedData<Invoice>>> {
  const res = await client.get<PaginatedData<Invoice>>(
    '/finance/invoices',
    params as Record<string, never>,
  )
  return res
}

export async function createInvoice(data: {
  invoice_code: string
  invoice_type?: string
  amount: number
  tax_amount?: number
  issued_at?: string
  issuer_name: string
  buyer_name?: string
  image_url?: string
}): Promise<ApiResponse<{ id: string }>> {
  const res = await client.post<{ id: string }>('/finance/invoices', data)
  return res
}

export async function updateInvoice(
  id: string,
  data: Partial<{
    invoice_code: string
    invoice_type: string
    amount: number
    tax_amount: number | null
    issued_at: string | null
    issuer_name: string
    buyer_name: string | null
    image_url: string | null
  }>,
): Promise<ApiResponse<null>> {
  const res = await client.put<null>(`/finance/invoices/${id}`, data)
  return res
}

export async function deleteInvoice(id: string): Promise<ApiResponse<null>> {
  const res = await client.delete<null>(`/finance/invoices/${id}`)
  return res
}

// ---- 收付款记录 ----

export interface PaymentQuery {
  page?: number
  page_size?: number
  direction?: string
  keyword?: string
  department_id?: string
  from?: string
  to?: string
}

export async function getPayments(
  params: PaymentQuery = {},
): Promise<ApiResponse<PaginatedData<Payment>>> {
  const res = await client.get<PaginatedData<Payment>>(
    '/finance/payments',
    params as Record<string, never>,
  )
  return res
}

export async function createPayment(data: {
  direction: 'income' | 'expense'
  category: string
  amount: number
  counterparty?: string
  occurred_at: string
  department_id?: string
  remark?: string
}): Promise<ApiResponse<{ id: string }>> {
  const res = await client.post<{ id: string }>('/finance/payments', data)
  return res
}

export async function updatePayment(
  id: string,
  data: Partial<{
    direction: 'income' | 'expense'
    category: string
    amount: number
    counterparty: string | null
    occurred_at: string
    department_id: string | null
    remark: string | null
  }>,
): Promise<ApiResponse<null>> {
  const res = await client.put<null>(`/finance/payments/${id}`, data)
  return res
}

export async function deletePayment(id: string): Promise<ApiResponse<null>> {
  const res = await client.delete<null>(`/finance/payments/${id}`)
  return res
}

// ---- 预算 ----

export interface BudgetQuery {
  page?: number
  page_size?: number
  period_type?: string
  period_value?: string
  department_id?: string
}

export async function getBudgets(
  params: BudgetQuery = {},
): Promise<ApiResponse<PaginatedData<Budget>>> {
  const res = await client.get<PaginatedData<Budget>>(
    '/finance/budgets',
    params as Record<string, never>,
  )
  return res
}

export async function createBudget(data: {
  department_id: string
  period_type: 'year' | 'month'
  period_value: string
  amount: number
}): Promise<ApiResponse<{ id: string }>> {
  const res = await client.post<{ id: string }>('/finance/budgets', data)
  return res
}

export async function updateBudget(
  id: string,
  data: Partial<{
    department_id: string
    period_type: string
    period_value: string
    amount: number
  }>,
): Promise<ApiResponse<null>> {
  const res = await client.put<null>(`/finance/budgets/${id}`, data)
  return res
}

export async function deleteBudget(id: string): Promise<ApiResponse<null>> {
  const res = await client.delete<null>(`/finance/budgets/${id}`)
  return res
}

// ---- 财务报表 ----

export interface ReportQuery {
  from?: string
  to?: string
  granularity?: string
}

export async function getReportSummary(
  params: ReportQuery = {},
): Promise<ApiResponse<ReportSummary>> {
  const res = await client.get<ReportSummary>(
    '/finance/reports/summary',
    params as Record<string, never>,
  )
  return res
}

export async function getReportDepartments(
  params: ReportQuery = {},
): Promise<ApiResponse<{ items: DepartmentReportRow[]; total_expense: number }>> {
  const res = await client.get<{ items: DepartmentReportRow[]; total_expense: number }>(
    '/finance/reports/departments',
    params as Record<string, never>,
  )
  return res
}

export async function getReportTrend(
  params: ReportQuery = {},
): Promise<ApiResponse<{ items: TrendRow[] }>> {
  const res = await client.get<{ items: TrendRow[] }>(
    '/finance/reports/trend',
    params as Record<string, never>,
  )
  return res
}

/** 打开 CSV 导出（浏览器直链下载，携带 Cookie 认证）。 */
export function exportReportUrl(
  type: 'reimbursements' | 'payments',
  params: ReportQuery = {},
): string {
  const search = new URLSearchParams()
  if (params.from) search.set('from', params.from)
  if (params.to) search.set('to', params.to)
  const qs = search.toString()
  return `/api/finance/reports/export/${type}${qs ? `?${qs}` : ''}`
}

File: backend/src/models/finance.rs
// Manner_Web - 可以在 Linux 系统上运行的企业管理系统
// Copyright (C) 2026 Linux-System-0(Github) / 一架在Linux上起飞的A320(Bilibili) <ls0_1@qq.com>
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

//! 财务模块数据模型：报销单 / 发票 / 收付款记录 / 预算。
//! 金额列在库中为 DECIMAL(12,2)，读取时经 `CAST(... AS DOUBLE)` 映射为 f64
//! （sqlx 不对 DECIMAL 直接解码 f64），写入时绑定 f64（MySQL 隐式转换）。

use chrono::NaiveDate;
use chrono::NaiveDateTime;
use serde::{Deserialize, Serialize};
use sqlx::FromRow;

// ---- 报销单 ----

/// 报销单列表行。
#[derive(Debug, Serialize, FromRow)]
pub struct ReimbursementListRow {
    pub id: String,
    pub employee_id: String,
    pub employee_name: String,
    pub department_id: String,
    pub department_name: String,
    pub title: String,
    pub category: String,
    pub amount: f64,
    pub currency: String,
    pub status: String,
    pub approver_id: Option<String>,
    pub approver_name: Option<String>,
    pub approve_comment: Option<String>,
    pub finance_reviewer_id: Option<String>,
    pub finance_reviewer_name: Option<String>,
    pub finance_comment: Option<String>,
    pub paid_at: Option<NaiveDateTime>,
    pub created_at: NaiveDateTime,
}

/// 报销单详情内部行（FromRow，供详情查询使用）。
#[derive(Debug, FromRow)]
pub struct ReimbursementDetailRow {
    pub employee_id: String,
    pub employee_name: String,
    pub department_id: String,
    pub department_name: String,
    pub title: String,
    pub category: String,
    pub amount: f64,
    pub currency: String,
    pub reason: Option<String>,
    pub status: String,
    pub approver_id: Option<String>,
    pub approver_name: Option<String>,
    pub approve_comment: Option<String>,
    pub approved_at: Option<NaiveDateTime>,
    pub finance_reviewer_id: Option<String>,
    pub finance_reviewer_name: Option<String>,
    pub finance_comment: Option<String>,
    pub finance_reviewed_at: Option<NaiveDateTime>,
    pub paid_at: Option<NaiveDateTime>,
    pub created_at: NaiveDateTime,
}

/// 报销单详情（含发票与审批流水）。
#[derive(Debug, Serialize)]
pub struct ReimbursementDetail {
    pub id: String,
    pub employee_id: String,
    pub employee_name: String,
    pub department_id: String,
    pub department_name: String,
    pub title: String,
    pub category: String,
    pub amount: f64,
    pub currency: String,
    pub reason: Option<String>,
    pub status: String,
    pub approver_id: Option<String>,
    pub approver_name: Option<String>,
    pub approve_comment: Option<String>,
    pub approved_at: Option<NaiveDateTime>,
    pub finance_reviewer_id: Option<String>,
    pub finance_reviewer_name: Option<String>,
    pub finance_comment: Option<String>,
    pub finance_reviewed_at: Option<NaiveDateTime>,
    pub paid_at: Option<NaiveDateTime>,
    pub created_at: NaiveDateTime,
    pub invoices: Vec<InvoiceRow>,
    pub logs: Vec<ReimbursementLogRow>,
}

/// 审批/状态流水行。
#[derive(Debug, Serialize, FromRow)]
pub struct ReimbursementLogRow {
    pub id: String,
    pub action: String,
    pub actor_id: String,
    pub actor_name: String,
    pub comment: Option<String>,
    pub created_at: NaiveDateTime,
}

/// 新建报销单请求。
#[derive(Debug, Deserialize)]
pub struct NewReimbursement {
    pub title: String,
    pub category: String,
    pub amount: f64,
    pub reason: Option<String>,
    /// 关联发票 id 列表（可空；invoice 须存在且未被其他报销单占用）。
    #[serde(default)]
    pub invoice_ids: Vec<String>,
}

/// 更新报销单请求（仅 pending_leader / rejected / withdrawn 状态可编辑并重新提交）。
#[derive(Debug, Deserialize)]
pub struct UpdateReimbursement {
    pub title: Option<String>,
    pub category: Option<String>,
    pub amount: Option<f64>,
    pub reason: Option<Option<String>>,
    #[serde(default)]
    pub invoice_ids: Option<Vec<String>>,
}

/// 审批/复核动作请求（approve 通过 / reject 驳回）。
#[derive(Debug, Deserialize)]
pub struct ReviewAction {
    pub action: String,
    pub comment: Option<String>,
}

/// 报销单查询参数。
#[derive(Debug, Deserialize)]
pub struct ReimbursementQuery {
    pub page: Option<i64>,
    pub page_size: Option<i64>,
    pub status: Option<String>,
    pub keyword: Option<String>,
    pub department_id: Option<String>,
}

// ---- 发票 ----

/// 发票行。
#[derive(Debug, Serialize, FromRow, Clone)]
pub struct InvoiceRow {
    pub id: String,
    pub invoice_code: String,
    pub invoice_type: String,
    pub amount: f64,
    pub tax_amount: Option<f64>,
    pub issued_at: Option<NaiveDate>,
    pub issuer_name: String,
    pub buyer_name: Option<String>,
    pub image_url: Option<String>,
    pub employee_id: String,
    pub employee_name: String,
    /// unused 未关联 | claimed 已关联报销单。
    pub status: String,
    pub created_at: NaiveDateTime,
}

/// 新建发票请求。
#[derive(Debug, Deserialize)]
pub struct NewInvoice {
    pub invoice_code: String,
    pub invoice_type: Option<String>,
    pub amount: f64,
    pub tax_amount: Option<f64>,
    pub issued_at: Option<NaiveDate>,
    pub issuer_name: String,
    pub buyer_name: Option<String>,
    pub image_url: Option<String>,
}

/// 更新发票请求。
#[derive(Debug, Default, Deserialize)]
#[serde(default)]
pub struct UpdateInvoice {
    pub invoice_code: Option<String>,
    pub invoice_type: Option<String>,
    pub amount: Option<f64>,
    pub tax_amount: Option<Option<f64>>,
    pub issued_at: Option<Option<NaiveDate>>,
    pub issuer_name: Option<String>,
    pub buyer_name: Option<Option<String>>,
    pub image_url: Option<Option<String>>,
}

/// 发票查询参数。
#[derive(Debug, Deserialize)]
pub struct InvoiceQuery {
    pub page: Option<i64>,
    pub page_size: Option<i64>,
    pub keyword: Option<String>,
    pub status: Option<String>,
}

// ---- 收付款记录 ----

/// 收付款记录行。
#[derive(Debug, Serialize, FromRow)]
pub struct PaymentRow {
    pub id: String,
    /// income 收款 | expense 付款。
    pub direction: String,
    pub category: String,
    pub amount: f64,
    pub counterparty: Option<String>,
    pub occurred_at: NaiveDate,
    pub department_id: Option<String>,
    pub department_name: Option<String>,
    pub remark: Option<String>,
    /// 关联报销单 id（报销付款自动生成时非空）。
    pub reimbursement_id: Option<String>,
    pub created_by: String,
    pub creator_name: String,
    pub created_at: NaiveDateTime,
}

/// 新建收付款记录请求。
#[derive(Debug, Deserialize)]
pub struct NewPayment {
    pub direction: String,
    pub category: String,
    pub amount: f64,
    pub counterparty: Option<String>,
    pub occurred_at: NaiveDate,
    pub department_id: Option<String>,
    pub remark: Option<String>,
}

/// 更新收付款记录请求。
#[derive(Debug, Default, Deserialize)]
#[serde(default)]
pub struct UpdatePayment {
    pub direction: Option<String>,
    pub category: Option<String>,
    pub amount: Option<f64>,
    pub counterparty: Option<Option<String>>,
    pub occurred_at: Option<NaiveDate>,
    pub department_id: Option<Option<String>>,
    pub remark: Option<Option<String>>,
}

/// 收付款查询参数。
#[derive(Debug, Deserialize)]
pub struct PaymentQuery {
    pub page: Option<i64>,
    pub page_size: Option<i64>,
    pub direction: Option<String>,
    pub keyword: Option<String>,
    pub department_id: Option<String>,
    pub from: Option<NaiveDate>,
    pub to: Option<NaiveDate>,
}

// ---- 预算 ----

/// 预算原始行（不含聚合列，spent/remaining 由 handler 计算）。
#[derive(Debug, FromRow)]
pub struct BudgetRawRow {
    pub id: String,
    pub department_id: String,
    pub department_name: String,
    pub period_type: String,
    pub period_value: String,
    pub amount: f64,
    pub created_at: NaiveDateTime,
}

/// 新建预算请求。
#[derive(Debug, Deserialize)]
pub struct NewBudget {
    pub department_id: String,
    pub period_type: String,
    pub period_value: String,
    pub amount: f64,
}

/// 更新预算请求。
#[derive(Debug, Default, Deserialize)]
#[serde(default)]
pub struct UpdateBudget {
    pub department_id: Option<String>,
    pub period_type: Option<String>,
    pub period_value: Option<String>,
    pub amount: Option<f64>,
}

/// 预算查询参数。
#[derive(Debug, Deserialize)]
pub struct BudgetQuery {
    pub page: Option<i64>,
    pub page_size: Option<i64>,
    pub period_type: Option<String>,
    pub period_value: Option<String>,
    pub department_id: Option<String>,
}

// ---- 财务报表 ----

/// 报表通用日期范围参数。
#[derive(Debug, Deserialize)]
pub struct ReportQuery {
    pub from: Option<NaiveDate>,
    pub to: Option<NaiveDate>,
    /// trend 粒度：month | year（默认 month）。
    pub granularity: Option<String>,
}

/// 汇总报表：期间内收/支/结余与报销统计。
#[derive(Debug, Serialize)]
pub struct ReportSummary {
    pub income: f64,
    pub expense: f64,
    pub net: f64,
    pub income_count: i64,
    pub expense_count: i64,
    pub reimbursement_pending: f64,
    pub reimbursement_pending_count: i64,
}

/// 部门费用排行行。
#[derive(Debug, Serialize)]
pub struct DepartmentReportRow {
    pub department_id: String,
    pub department_name: String,
    pub expense: f64,
}

/// 趋势行（按月份或年份聚合收/支）。
#[derive(Debug, Serialize)]
pub struct TrendRow {
    pub period: String,
    pub income: f64,
    pub expense: f64,
}
